# Special Pythagorean Triplet

## Problem Statement -

A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
$$а^2 + b^2=c^2$$
For example, $3^2 + 4^2 = 9 + 16 = 25 = 52$
There exists exactly one Pythagorean triplet for which $a + b + c = 1000.$
Find the product abc.

## Answers -

## Vini's-logic (human, some minutes)

- Three squared numbers must be summed to reach 1000;
- Just for fun, let's approximate it

In [92]:
import numpy as np


In [93]:
three = np.random.rand(3)
total = 0
difference = 0
while total < 999.9:
    total = sum(three**2)
    difference = 1000 - total
    three += 2 * three * (difference / 1000) * 0.01
print(f"Total: {total:.2f}, Difference: {difference:.2f}, Three: {three}")

Total: 999.90, Difference: 0.10, Three: [19.8592128  21.03770219 12.76458273]


- Cool function approximation, but does not help us
- Another method is needed
- $32$ is the maximum before passing $1000$. $32\cdot 32 = 1024$
- after carefully inspecting the equations

In [94]:
for a in range(1, 500):  
    for b in range(a + 1, 1000 - a):  
        c = 1000 - a - b  
        if a**2 + b**2 == c**2:  
            print(f"The result abc = {a * b * c}")
            break  

The result abc = 31875000


# O3-mini (11 seconds)

# Pythagorean Triplet Problem

A *Pythagorean triplet* is a set of three natural numbers, $a < b < c$, that satisfy the equation:

$$a^2 + b^2 = c^2.$$

We are looking for the unique triplet where

$$a + b + c = 1000.$$

For example, the triplet $ (3, 4, 5) $ is Pythagorean because:

$$3^2 + 4^2 = 9 + 16 = 25 = 5^2.$$

The following code iterates over possible values of $a$ and $b$, computes $c = 1000 - a - b$, and checks the condition above. Once the triplet is found, it prints the values and the product $$abc$$.


In [95]:
# Function to find the Pythagorean triplet for which a + b + c equals the given sum.
def find_pythagorean_triplet(total_sum):
    for a in range(1, total_sum):
        for b in range(a + 1, total_sum - a):
            c = total_sum - a - b
            if a**2 + b**2 == c**2 and c > b:
                return a, b, c
    return None, None, None

# Given sum
total_sum = 1000

# Find the triplet and calculate the product
a, b, c = find_pythagorean_triplet(total_sum)
if a and b and c:
    product = a * b * c
    print(f"Pythagorean triplet: a = {a}, b = {b}, c = {c}")
    print(f"Product abc: {product}")
else:
    print("No triplet found that satisfies the conditions.")


Pythagorean triplet: a = 200, b = 375, c = 425
Product abc: 31875000


## DeepSeek-R1 (235 seconds)

### Step-by-Step Explanation of the Pythagorean Triplet Problem

### 1. Understand the Problem
We need three natural numbers $a < b < c$ such that
$$a^2 + b^2 = c^2,$$
and their sum is $1000$. The goal is to find their product $abc$.

Given:
$a + b + c = 1000,$
we can express $c$ as:
$c = 1000 - a - b.$

Substitute $c = 1000 - a - b$ into the Pythagorean equation:
$$a^2 + b^2 = (1000 - a - b)^2.$$

Expanding and simplifying leads to the equation:
$ab - 1000a - 1000b + 500000 = 0.$
This can be factorized into:
$(a - 1000)(b - 1000) = 500000.$

Find pairs of factors of $500000$ where both factors are less than $1000$ and maintain the condition $a < b < c$.

The valid factor pair is:
$d = 800 \quad \text{and} \quad e = 625,$
which leads to the triplet:
$a = 200,\quad b = 375,\quad c = 425.$

Compute the product:
$$abc = 200 \times 375 \times 425 = 31875000.$$


In [96]:
for k in range(500, 1000):
    if 500000 % k == 0:
        m = 500000 // k
        if m < k:
            a = 1000 - k
            b = 1000 - m
            c = 1000 - a - b
            if a > 0 and b > 0 and c > 0 and a < b:
                if a**2 + b**2 == c**2:
                    print(f"Triplet found: a={a}, b={b}, c={c}")
                    print(f"Product abc = {a * b * c}")
                    break

Triplet found: a=200, b=375, c=425
Product abc = 31875000
